File: src/main/java/com/github/onlaait/mcserverpinger/ServerAddress.java
package com.github.onlaait.mcserverpinger;

import com.google.common.net.HostAndPort;

import java.net.IDN;

public final class ServerAddress {
    private final HostAndPort hostAndPort;

    public ServerAddress(String host, int port) {
        this(HostAndPort.fromParts(host, port));
    }

    private ServerAddress(HostAndPort hostAndPort) {
        this.hostAndPort = hostAndPort;
    }

    public String getAddress() {
        try {
            return IDN.toASCII(this.hostAndPort.getHost());
        } catch (IllegalArgumentException illegalArgumentException) {
            return "";
        }
    }

    public int getPort() {
        return this.hostAndPort.getPort();
    }

    public static ServerAddress parse(String address) throws PingerException {
        if (address == null) {
            throw new PingerException("com.github.onlaait.mcserverpinger.Address is null");
        }
        try {
            HostAndPort hostAndPort = HostAndPort.fromString(address).withDefaultPort(25565);
            if (hostAndPort.getHost().isEmpty()) {
                throw new PingerException("Host is empty");
            }
            return new ServerAddress(hostAndPort);
        } catch (IllegalArgumentException illegalArgumentException) {
            throw new PingerException("Failed to parse URL " + address);
        }
    }

    static int portOrDefault(String port) {
        try {
            return Integer.parseInt(port.trim());
        } catch (Exception exception) {
            return 25565;
        }
    }

    public String toString() {
        return this.hostAndPort.toString();
    }

    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o instanceof ServerAddress) {
            return this.hostAndPort.equals(((ServerAddress)o).hostAndPort);
        }
        return false;
    }

    public int hashCode() {
        return this.hostAndPort.hashCode();
    }
}

File: src/main/kotlin/com/github/onlaait/mcserverpinger/address/ServerAddress.kt
package com.github.onlaait.mcserverpinger.address

import com.github.onlaait.mcserverpinger.PingerException
import com.google.common.net.HostAndPort

import java.net.IDN

class ServerAddress {

    constructor(host: String, port: Int) : this(HostAndPort.fromParts(host, port))

    private constructor(hostAndPort: HostAndPort) {
        this.hostAndPort = hostAndPort
    }

    companion object {
        fun parse(address: String): ServerAddress =
            try {
                val hostAndPort = HostAndPort.fromString(address).withDefaultPort(25565)
                if (hostAndPort.host.isEmpty()) throw PingerException("Host is empty")
                ServerAddress(hostAndPort)
            } catch (_: IllegalArgumentException) {
                throw PingerException("Failed to parse URL $address")
            }

        fun portOrDefault(port: String): Int =
            try {
                port.trim().toInt()
            } catch (_: Exception) {
                25565
            }
    }

    private val hostAndPort: HostAndPort

    val address: String
        get() =
            try {
                IDN.toASCII(hostAndPort.host)
            } catch (_: IllegalArgumentException) {
                ""
            }

    val port: Int
        get() = hostAndPort.port

    override fun toString(): String = hostAndPort.toString()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other is ServerAddress) return hostAndPort == other.hostAndPort
        return false
    }

    override fun hashCode(): Int = hostAndPort.hashCode()
}